pkg/operator/certrotationcontroller/satokensigner_controller_test.go

```diff
@@ -0,0 +1,104 @@
+package certrotationcontroller
+
+import "testing"
+
+func TestIsValidRSAKeyPair(t *testing.T) {
+	var (
+		rsaPublicKey1 = []byte(`
+-----BEGIN PUBLIC KEY-----
+MIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQCqGKukO1De7zhZj6+H0qtjTkVxwTCpvKe4eCZ0
+FPqri0cb2JZfXJ/DgYSF6vUpwmJG8wVQZKjeGcjDOL5UlsuusFncCzWBQ7RKNUSesmQRMSGkVb1/
+3j+skZ6UtW+5u09lHNsj6tQ51s1SPrCBkedbNf0Tp0GbMJDyR4e9T04ZZwIDAQAB
+-----END PUBLIC KEY-----
+`)
+
+		rsaPrivKey1 = []byte(`
+-----BEGIN RSA PRIVATE KEY-----
+MIICXAIBAAKBgQCqGKukO1De7zhZj6+H0qtjTkVxwTCpvKe4eCZ0FPqri0cb2JZfXJ/DgYSF6vUp
+wmJG8wVQZKjeGcjDOL5UlsuusFncCzWBQ7RKNUSesmQRMSGkVb1/3j+skZ6UtW+5u09lHNsj6tQ5
+1s1SPrCBkedbNf0Tp0GbMJDyR4e9T04ZZwIDAQABAoGAFijko56+qGyN8M0RVyaRAXz++xTqHBLh
+3tx4VgMtrQ+WEgCjhoTwo23KMBAuJGSYnRmoBZM3lMfTKevIkAidPExvYCdm5dYq3XToLkkLv5L2
+pIIVOFMDG+KESnAFV7l2c+cnzRMW0+b6f8mR1CJzZuxVLL6Q02fvLi55/mbSYxECQQDeAw6fiIQX
+GukBI4eMZZt4nscy2o12KyYner3VpoeE+Np2q+Z3pvAMd/aNzQ/W9WaI+NRfcxUJrmfPwIGm63il
+AkEAxCL5HQb2bQr4ByorcMWm/hEP2MZzROV73yF41hPsRC9m66KrheO9HPTJuo3/9s5p+sqGxOlF
+L0NDt4SkosjgGwJAFklyR1uZ/wPJjj611cdBcztlPdqoxssQGnh85BzCj/u3WqBpE2vjvyyvyI5k
+X6zk7S0ljKtt2jny2+00VsBerQJBAJGC1Mg5Oydo5NwD6BiROrPxGo2bpTbu/fhrT8ebHkTz2epl
+U9VQQSQzY1oZMVX8i1m5WUTLPz2yLJIBQVdXqhMCQBGoiuSoSjafUhV7i1cEGpb88h5NBYZzWXGZ
+37sJ5QsW+sJyoNde3xH8vdXhzU7eT82D6X/scw9RZz+/6rCJ4p0=
+-----END RSA PRIVATE KEY-----
+`)
+
+		rsaPublicKey2 = []byte(`
+-----BEGIN PUBLIC KEY-----
+MIGeMA0GCSqGSIb3DQEBAQUAA4GMADCBiAKBgG3nScV2wLxaS3JaEHJrepzbXmql
+nh0BDYdr4GRjVR6EeC1E0edO1LiiwI/aU7xbXa0wHEI4kr/MnRDIlV+7L/6FLqob
+PH8fg5HM0K2dE2vaEWIb8saRWs8r49tqeChiHsPEGeJeofKgeXw0XrEW6+l7QQO5
+vH+y1RdSulDn33KlAgMBAAE=
+-----END PUBLIC KEY-----
+`)
+
+		multipleKeys = []byte(`
+-----BEGIN PUBLIC KEY-----
+MIGeMA0GCSqGSIb3DQEBAQUAA4GMADCBiAKBgHp2orP6kNoVtYMW9l4Q0p9iTcFV
+OsFOyuHs7qNreZql4CKa6qReyk9svM+XvGnzgqpmCN/CjeMx1t13gLn0ZUeKK9Sv
+v/n5xutjIEtKz1DRZRsR4R3/+X9JGGkLdJdopAWol5dkrNIskASEmWWN7R55eXmg
+HeOzuVBZV9Nq+N0XAgMBAAE=
+-----END PUBLIC KEY-----
+-----BEGIN PUBLIC KEY-----
+MIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQDAh14+qIRu+CdE6wlyg4WMsc3j
+80W5sbZccH4dPxoEGlWMa8B2A+olOAy5qw8KoU3Xl1yuND8QvB3Xb499GGIX0aqN
+OTVwaSKxTZDSGnoJipZsxhxzDpHi6rn/pAdE4jnkqfaqujZbnTyHRhNdvy7jVO7d
+s16gDilgo+8DEAxQfQIDAQAB
+-----END PUBLIC KEY-----
+`)
+	)
+
+	var keyPairTests = []struct {
+		name    string
+		pubKey  []byte
+		privKey []byte
+		exp     bool
+	}{
+		{
+			name:    "matching keys",
+			pubKey:  rsaPublicKey1,
+			privKey: rsaPrivKey1,
+			exp:     true,
+		},
+		{
+			name:    "not matching keys",
+			pubKey:  rsaPublicKey2,
+			privKey: rsaPrivKey1,
+			exp:     false,
+		},
+		{
+			name:    "allow only one key",
+			pubKey:  multipleKeys,
```

Comment:
Is this genuinely an error? If you're not sure, please check the code in the apiserver that validates tokens issued by the private key.

Reply:
Purpose of `isValidKeyPair` is given one public key and one private key, check if both the keys match i.e. can tokens signed with private key be validated  using public key. It may not be able to do this if multiple keys are passed as part of arguments. Right now it is outside the context of this function to check if multiple keys are passed as arguments. And I might need more context if this is what we want.

I checked  controller-manager code to see how it parses keys and ended up looking at the `keyutil` package and all its APIs to understand how is it implement. As far as I understand, it is slightly different from what this function is doing. Here we are to be only concerned if a key pair matches. In the code for controller-manager or apiserver, there is a possibility that we might have multiple keys. two different objectives?

Reply:
The apiserver supports multiple public keys - this allows validation of tokens issued by previous private key as well as the current one. The controller manager only supports a single keypair. My question is whether the controller manager will error out if the public or private key entries in the signing secret contain more than one key, or if it will simply pick the first of many keys supplied.

Reply:
I updated the implementation now to follow whatever `kube-controller-manager` and `kube-apiserver` are doing. please take a look